### DISCUSSION TITLE
Request - Bring back delete functionality in Client

### DISCUSSION THREAD

**Discussion by BB-Forrest** [jellyfin/jellyfin-androidtv#2368] [General]
In my opinion (and only mine!) various clients for Jellyfin should have the same functionality.

With the release of v0.15 (great job by the way) we find ourselves in the position of watching a program and then if we want to delete it we can use the IOS version of the app to do so.

This seems a little counter intuitive to me.

Perhaps a design compromise could be that Admin users have delete privileges through the app but normal clients don't?

I understand the argument that a client shouldn't be managing a server but I would argue that an admin client user should be able to manage the media on the server - it's not as though they can create and delete library's, add plug ins or change transcoding settings etc through it.

Any thoughts appreciated.

**Comment by quielb** [comment]
I don't consider the delete function as part of "managing" the server. I consider it to be part of the flow of watching a show. If I watch something and want to then delete it because I have no intention of doing so again, I should be able to delete it from the front-end. In fact I would like the delete function to appear on more levels of the menu. For example, having being able to delete a season of a show as the new season starts. Or in the movies thumbnail view. Having to drill in to the actual movie to delete requires extra clicks that I would prefer to not do.

**Comment by MailYouLater** [comment]
Once I get my Jellyfin server fully set up, the Android TV app will be the _only_ way my family members are likely to _ever_ be interacting with it.  Mostly they will be recording live TV and watching it later, but occasionally I will rip movies/episodes from optical media for them to watch.

As I have limited storage on my server, I need them to be able to delete anything they're not planning to watch again, because I don't want to have to go around asking everybody what I can delete (without upsetting anyone) to make space to rip another set of discs to the server.

**Comment by gassssss** [comment]
In my opinion removing the option is a step backwards in usability. I mostly consume my media on an Android TV and having to login on my phone or PC is a slight annoyance, but of course I also understand the reasoning behind it. 

Maybe a more accommodating alternative would be making it a non default option so if you want to have it you have to specifically enable it. Another way to implement it in order to encourage making use of multi-user feature in Jellyfin could be a "mark for deletion" flag that would delete the file only once a set percentage of the users or an administrator has set that flag.

I don't know the best way, I'm just throwing some ideas to perhaps consider, but I do know that I miss the bug/feature. 

Needless to say I really appreciate the hard work on fantastic software and I am really grateful for it.

**Reply by MailYouLater** [reply]
I really like the idea of an option for users without deletion privileges to be able to mark a piece of media for removal, and a corresponding option for a user with deletion privileges to be able to approve it so that the system will actually delete it.  However... I have a feeling adding this kind of functionality would be a bit more in depth than simply restoring the delete button, and would probably require some degree of implementation on both the client and server sides.

I really do hope that something like that does get implemented at some point because I feel like a great [semi]automated media removal system could be built around it, where a user can mark something for removal, another user can challenge the removal request, an administrator can look at the removal queue to see how many/which users voted each way to guide their decision to approve/deny the removal request, and/or a scheduled (e.g. weekly/monthly) task could delete any media that was marked for deletion and hasn't been challenged.  Such a media removal system would be perfect for my use case.

In the meantime however, I'd be really happy just to have the delete option that according to what I'm reading, _was_ present, just a few days before I installed Jellyfin to make sure it would work for my needs.

**Comment by smstarrant** [comment]
I consume all of my media through the Android TV client, and am the only user, so no issue with accidental deletions.  It just makes more sense to me that the option to remove episodes be in the place I just watched them.

**Comment by MailYouLater** [comment]
It should also be noted that there are already permissions for allowing/disallowing deletion on a per-user basis when creating/modifying user accounts on a Jellyfin server, so if you don't trust your users not to delete your media, then just don't give them deletion permissions when you make an account for them.

Here's a screenshot:
![Allow media deletion from:
☐ All libraries
☑ Recordings
☐ Movies
☑ Shows](https://user-images.githubusercontent.com/1279475/210147858-a29afeaf-740e-465b-a227-1b67246b2d32.png)

**Comment by BB-Forrest** [comment]
I guess my view is that all Jellyfin clients should have a core functionality that is defined and consistent across all clients as part of the core project, obviously the look and feel of individual clients is going to be different.

Any change to that core functionality should be agreed across the entire project - i.e. If the client cannot delete media then every version of the client shouldn't be able to delete media, inconsistencies between client versions lead to problems down the line (I'm a senior techy for a multinational - not a coder sorry, more a jack of all trades while being a master of none - and have seen this at first hand on many systems over the years).

Happy to have a conversation about it with the project leads.

Happy New Year everyone.

**Comment by erbmur** [comment]
I believe there should be a delete button on the Android TV app. Me and my family consume 90% or more of our media through the TV app and losing the ability to delete shows after watching directly through the TV reduces from the user experience. This means we have to go to a second device once we've watched anything to be able to delete it.

**Reply by thomcuddihy** [reply]
Honestly thought the removal in 15 was a bug, and waited until the next release 0.15.1 for it to be fixed. Finding that not to be the case, I looked into it and found both the locked PR for it, and this thread. I respectfully disagree with @nielsvanvelzen's position entirely, and so does every other commenter so far. If that's not enough reason to rollback #1936 then I don't know what is.

**Comment by alpersoft** [comment]
I also think that the delete functionality should be restored, as on the server you can configure for each user if he can or cannot delete something from the library.

**Comment by Euromancer** [comment]
Im missing the delete button aswell. When im done watching a movie or show, then i delete it, because i have no more use for it. And having to do it in the android app afterwards, seems like a silly workaround...
Besides if someone are scared of users deleting files by mistanke, then just disable delete media the option on the user preferences page.

**Comment by digdilem** [comment]
Please reinstate this - I was shocked to discover it had been removed.

I delete TV and films once I've watched them through this client - removing this now requires me to remember what I've watched and launch a web client to remove it. That's a horrible user experience and really detracts from this version. 

Although I've read the reasoning for removing this function, I disagree strongly with it. It's rare for a project to remove functionality seemingly from an arbitrary decision especially when it cripples an important usability aspect like this.

**Reply by MailYouLater** [reply]
... Unfortunately it's not as rare as you might think, especially in projects with a small development team <sub><sup>(like how this one is currently driven mostly by _one person_)</sup></sub> but even if it were literally an every day occurance, that doesn't make it right.

Thankfully however, that person in this project listened to the users, and was willing to add the feature back. 🎉

**Comment by nielsvanvelzen** [comment]
Thank you everyone for your feedback. The delete option will be added back with #2415. It's a new implementation that fixes the issues with the previous one.

**Reply by digdilem** [reply]
Thank you, this is very good news!

**Reply by BB-Forrest** [reply]
Thanks Niels, appreciate your efforts.

**Reply by MailYouLater** [reply]
Awesome.  I look forward to seeing the new & improved delete function in action.

**Reply by Kithkani** [reply]
Thank you so much for adding this back.

**Reply by ikus060** [reply]
This is great ! Was coming here to understand what going on with the delete button.

Any release date for the fix ?

**Reply by nielsvanvelzen** [reply]
We don't do ETA's. The feature will be available in the next minor version (0.16) which I'm quite sure will be April or later.

**Reply by Marc3l1988** [reply]
i moved from Emby to Jellyfin since Jellyfin is opensource. But the delete function is one of the things that i use everytime.
When i finished watching something i like to delete it since i dont have much hdd space.

Is there something you can do with a quickfix, or something that the users can adjust in the code of the server?
(I'm using the Android TV application) When i am on Windows browser i can delete without any problems.

**Reply by nielsvanvelzen** [reply]
No, the feature is not backportable

**Comment by icrn** [comment]
hello，这里严康的邮箱，我已经收到了！

**Comment by ks0002544** [comment]
So I deleted v0.15.3 and installed v0.14.5. Function Delete must be in 0.14.x

**Reply by Marc3l1988** [reply]
So in V0.14.5 there is the delete function?

**Reply by ks0002544** [reply]
yes delete funtion is alive

**Reply by Marc3l1988** [reply]
To bad that you are not looking into the Android TV version one.
That is the one that has no delete function.

**Reply by MailYouLater** [reply]
@Marc3l1988: On the contrary; as stated by **ks0002544,** media deletion functionality was present in the 0.14.x releases of the Android TV client for Jellyfin.  It was removed in the first 0.15.x release because it apparently had some bugs, and **nielsvanvelzen** didn't think it was necessary or worth fixing.  After several users spoke out in favor of the feature, **nielsvanvelzen** took the time to reimplement the feature, and it is expected to return (enhanced, and hopefully bug-free) in the upcoming 0.16.x releases.

**Reply by ks0002544** [reply]
if version 14.5 has been installed it is necessary to switch to external player as vlc otherwise exoplayer ends Jellyfin with crash

**Comment by ks0002544** [comment]
nielsvanvelzen  
if you plan to bring back the Delete feature, please consider deleting the entire series as well. Previously, you could only delete one episode

**Comment by oscarchd** [comment]
feel that we should be able to delete the file in app

eg. once we finished an episode or the whole series, we can delete them straight from the app and dont need to go back to the server side to delete

really wish we can bring back the delete button

**Reply by MailYouLater** [reply]
This feature has been rewritten to fix the issues it had in 0.14.x and should be back in the 0.16.x release cycle.

**Reply by brianjmurrell** [reply]
Is there even a ballpark ETA on 0.16.x with this functionality?  This functionality is sorely missed.  I am now constantly being interrupted from what I am doing to delete shows for the rest of the family.  It's silly.

**Reply by nielsvanvelzen** [reply]
> Is there even a ballpark ETA on 0.16.x with this functionality?

From my other comment:

> We don't do ETA's. The feature will be available in the next minor version (0.16) which I'm quite sure will be April or later.

And April feels unlikely right now

**Reply by brianjmurrell** [reply]
So "April feels unlikely" means at least 2 more months of my family running back and forth telling me what they have watched and me having to log into JF just to delete shows every time.  ~sigh~

Is this something that cannot easily it info a 0.15.x bugfix release then where it would be much smaller and easier to QA.

**Reply by Marc3l1988** [reply]
When i read: And April feels unlikely right now i will be leaving Jellyfin for what it is and installed Plex for me and my familly.
Sorry to go, but a option to delete a seen movie/serie that is a must have.

**Reply by brianjmurrell** [reply]
I cannot disagree with the level of importance that @Marc3l1988 is expressing for this functionality.  It's crazy to me that it was even considered to be removed in the first place.

**Reply by neuroid** [reply]
Gents, this is an open source project, maintained by volunteers in their spare time, who I'm sure also have families and other commitments.
Your frustration is understandable - I was bemused too. But props to @nielsvanvelzen for re-adding the feature promptly.
I'm sure there are good reasons behind the current release cycle. I also suspect that complaining here will have much less impact than contributing or donating to the project.

**Reply by digdilem** [reply]
Agree with @neuroid - we must keep a sense of perspective and check our frustrations. 

Being a developer, especially of a free project, is a thankless task and regardless of this relatively minor, issue, we owe @nielsvanvelzen our appreciation for all they have done on this project.  Please be patient, I'm sure the fix will be available as soon as is possible.

**Reply by BB-Forrest** [reply]
Correct, Niels is doing this work in his spare time, it's not his job, I'm extremely grateful for his efforts.

I understand and share some peoples frustrations (after all, I raised the topic) but stamping your foot and acting like an entitled ass is not useful to making people feel good about their work.

If you want to criticise and get involved in the project - great, brilliant, fantastic, I'm sure that people will be glad of the help.

If you just want to criticise without any accountability or involvement then with the greatest of love and respect - f*ck off.

**Reply by Marc3l1988** [reply]
> Correct, Niels is doing this work in his spare time, it's not his job, I'm extremely grateful for his efforts.
> 
> I understand and share some peoples frustrations (after all, I raised the topic) but stamping your foot and acting like an entitled ass is not useful to making people feel good about their work.
> 
> If you want to criticise and get involved in the project - great, brilliant, fantastic, I'm sure that people will be glad of the help.
> 
> If you just want to criticise without any accountability or involvement then with the greatest of love and respect - f*ck off.

i'm not acting like a entitled ass, and i do love the work for this project. But for me personally the delete option is a must have that's the only thing that i'm saying. I hope that there will be a solution in a next release, but it is not something i can wait on unfortunately. i'm glad to change back to Jellyfin in time when this is fixed. But for the meantime it's not working for me.

**Reply by nielsvanvelzen** [reply]
Thank you @neuroid, @digdilem and @BB-Forrest for your messages. Let me explain the situation so hopefully everyone can live on in peace.

We've discussed the "delete media" feature within the Jellyfin team a bunch of time and the server team concluded that they'd like to remove the feature because the current implementation is not optimal and they have no intention to maintain it. Because of this conclusion we decided to remove the feature in the Android TV app when I was refactoring a bunch of code to make it more maintainable. I had to choose to either rewrite or remove a feature that I, at the time, thought was not used a lot and was planned to be removed from the server anyway.
I chose to remove it, because I only have so much time to spend on the app and as we all know, there is A LOT of work that needs to happen, like the outdated UI or all the playback issues.

Nobody complained during the beta cycle of this release, only after it was released as complaints issues started to slowly arise. When people (in this thread) started to explain their use-cases I decided they were right, and as long as the server supports it we should re-add the delete option. I spend multiple hours rewriting that code from scratch to fix the issues in the previous implementation (previously the delete media permission wasn't read correctly when only enabling it for specific libraries) and making it work with the refactored code. The result of that will be in the 0.16 release because it was merged a few weeks ago.

We cannot backport these changes because they contain translation changes (due to a limitation of our infrastructure) and are built on top of code parts that changed after the 0.15 release, which are not ready for release. That's why the feature will be back next release. The next release will take some time though because I'm the only person working on this huge app while working a fulltime job that pays my bills and the intended changes for 0.16 are far from ready. Please understand that Jellyfin is a volunteer based project, we're not a for-profit company that pays our developers, I only get about 25 dollars a month from GitHub sponsors.

If this thread starts to escalate again I'll lock it because I'm getting tired ot these hateful comments from a few specific persons.
edit: well it's locked now

**Comment by ks0002544** [comment]
May be it will help to wait untill delete function will be back. I installed Jellyfin Android 2.4.4 on my Nvidia Shield. Works fine. Delete function is alive. Only I use ordinary pc mouse for navigation. 
or
Open in your internet browser of Android tv box: ip_Jellyfin_server:8096
Use mouse to navigate. Add this ip address to bookmark.

**Comment by brianjmurrell** [comment]
> getting tired ot these hateful comments from a few specific persons

I haven't read every message in this whole thread, but I doubt anyone is actually being _hateful_ towards you.  I certainly would like to think that is not the case.  I hope nobody is even feeling anything at all negative towards anyone in particular.

I suspect some people are just trying to express how important this functionality is.

> We've discussed the "delete media" feature within the Jellyfin team a bunch of time and the server team concluded that they'd like to remove the feature

Uhm.  That's an even more serious issue!  How exactly would anything ever get removed from JF is this happened?  Is there a ticket or discussion ongoing anywhere about this, so that we can head that off and not have a repeat of what happened here there?
